### ISSUE TITLE
Server side cancel Timer I fired for TCP

### ISSUE THREAD

**Issue by SylvainFontaineJive** [emiago/sipgo#148]
Hi!  Thanks you for such a good project.

I am implementing a SIP server and I am trying to get notified when an INVITE is canceled.
The call flow is the following:
```
10:00:14.560420   │        INVITE (SDP)         │
      +0.000014   │ ──────────────────────────> │
10:00:14.560665   │         100 Trying          │
      +0.000011   │ <────────────────────────── │
10:00:14.560770   │           CANCEL            │
      +0.000010   │ ──────────────────────────> │
10:00:14.560948   │   487 Request Terminated    │
      +0.000012   │ <────────────────────────── │
10:00:14.561009   │           200 OK            │
      +0.000010   │ <────────────────────────── │
10:00:14.561049   │             ACK             │
      +0.000010   │ ──────────────────────────> │
````
I need to cancel my background go routine running in background regarding this INVITE request.
How do I do this now? 

I tried waiting on `<-tx.Done()` but it never gets called.
Even if I see logs like:
```
{"level":"debug","caller":"transactionLayer","time":"2024-10-29T10:14:49-04:00","message":"Server transaction terminating"}
```

This piece of code seems to work
```
select {
	case m := <-tx.Acks(): // Handle ACK for response . ACKs on 2xx are send as different request
		log.Info("Got ACK", "m", m, )
		err := tx.Err()
		if errors.Is(err, sip.ErrTransactionCanceled) {
			// Transaction was canceled
			log.Info("Transaction canceled", "tx", tx)
			return
		}

```

But doesn't seem appropriate as it's called only at ACK, and I need to take action at CANCEL.

How can I hook on Cancel requests? 

Thanks,
Sylvain

**Comment by emiago** [comment]
Hi @SylvainFontaineJive . This indeed could be some issue. It was some of recent changes. 
By default yes, processing error should be after tx.Done() 

It would be great help maybe to run with `sip.TransactionFSMDebug = true` and have zerolog debug level. Just to confirm why this happens.

**Comment by SylvainFontaineJive** [comment]
Hello @emiago ,

Here are the logs
```json
{"level":"debug","caller":"transport<UDP>","time":"2024-10-30T08:50:14-04:00","message":"begin listening on UDP [::]:5060"}
{"level":"debug","time":"2024-10-30T08:50:26-04:00","message":"UDP read from [::]:5060 <- 127.0.0.1:5061:\nINVITE sip:service@127.0.0.1:5060 SIP/2.0\r\nVia: SIP/2.0/UDP 127.0.0.1:5061;branch=z9hG4bK4da8.26650\r\nFrom: sipp <sip:sipp@127.0.0.1:5061>;tag=26650SIPpTag001\r\nTo: service <sip:service@127.0.0.1:5060>\r\nCall-ID: 1-26650@127.0.0.1\r\nCSeq: 1 INVITE\r\nContact: sip:sipp@127.0.0.1:5061\r\nX-CustomHeader: Test; param1=1; param2=2\r\nMax-Forwards: 70\r\nSubject: Performance Test\r\nContent-Type: application/sdp\r\nContent-Length:   129\r\n\r\nv=0\r\no=user1 53655765 2353687637 IN IP4 127.0.0.1\r\ns=-\r\nc=IN IP4 127.0.0.1\r\nt=0 0\r\nm=audio 6000 RTP/AVP 0\r\na=rtpmap:0 PCMU/8000\r\n"}
{"level":"debug","caller":"transportlayer","network":"udp","addr":"127.0.0.1:5061","time":"2024-10-30T08:50:26-04:00","message":"getting connection"}
{"level":"debug","caller":"transactionLayer","tx":"z9hG4bK4da8.26650__127.0.0.1__5061__INVITE","time":"2024-10-30T08:50:26-04:00","message":"Server transaction initialized"}
{"level":"debug","caller":"transactionLayer","key":"z9hG4bK4da8.26650__127.0.0.1__5061__INVITE","input":"server_input_user_1xx","state":"inviteStateProcceeding-fm","time":"2024-10-30T08:50:26-04:00","message":"Changing transaction state"}
{"level":"debug","time":"2024-10-30T08:50:26-04:00","message":"UDP write to [::]:5060 -> 127.0.0.1:5061:\nSIP/2.0 100 Trying\r\nVia: SIP/2.0/UDP 127.0.0.1:5061;branch=z9hG4bK4da8.26650\r\nFrom: \"sipp\" <sip:sipp@127.0.0.1:5061>;tag=26650SIPpTag001\r\nTo: \"service\" <sip:service@127.0.0.1:5060>\r\nCall-ID: 1-26650@127.0.0.1\r\nCSeq: 1 INVITE\r\nContent-Length: 0\r\n\r\n"}
{"time":"2024-10-30T08:50:26.692368-04:00","level":"INFO","msg":"Received INVITE","request":{"SipVersion":"SIP/2.0","Method":"INVITE","Recipient":{"Scheme":"sip","Wildcard":false,"HierarhicalSlashes":false,"User":"service","Password":"","Host":"127.0.0.1","Port":5060,"UriParams":null,"Headers":null}}}
{"level":"debug","time":"2024-10-30T08:50:26-04:00","message":"UDP read from [::]:5060 <- 127.0.0.1:5061:\nCANCEL sip:service@127.0.0.1:5060 SIP/2.0\r\nVia: SIP/2.0/UDP 127.0.0.1:5061;branch=z9hG4bK4da8.26650\r\nFrom: sipp <sip:sipp@127.0.0.1:5061>;tag=26650SIPpTag001\r\nTo: service <sip:service@127.0.0.1:5060>\r\nCall-ID: 1-26650@127.0.0.1\r\nCSeq: 1 CANCEL\r\nContact: sip:sipp@127.0.0.1:5061\r\nMax-Forwards: 70\r\nSubject: Performance Test\r\nContent-Type: application/sdp\r\nContent-Length:     0\r\n\r\n"}
{"level":"debug","caller":"transactionLayer","key":"z9hG4bK4da8.26650__127.0.0.1__5061__INVITE","input":"server_input_cancel","state":"inviteStateProcceeding-fm","time":"2024-10-30T08:50:26-04:00","message":"Changing transaction state"}
{"level":"debug","caller":"transactionLayer","time":"2024-10-30T08:50:26-04:00","message":"Passing 487 on CANCEL"}
{"level":"debug","caller":"transactionLayer","key":"z9hG4bK4da8.26650__127.0.0.1__5061__INVITE","input":"server_input_user_300_plus","state":"inviteStateProcceeding-fm","time":"2024-10-30T08:50:26-04:00","message":"Changing transaction state"}
{"level":"debug","time":"2024-10-30T08:50:26-04:00","message":"UDP write to [::]:5060 -> 127.0.0.1:5061:\nSIP/2.0 487 Request Terminated\r\nVia: SIP/2.0/UDP 127.0.0.1:5061;branch=z9hG4bK4da8.26650\r\nFrom: \"sipp\" <sip:sipp@127.0.0.1:5061>;tag=26650SIPpTag001\r\nTo: \"service\" <sip:service@127.0.0.1:5060>;tag=8f4affcf-9ca4-4ebe-a36b-4f60f844509b\r\nCall-ID: 1-26650@127.0.0.1\r\nCSeq: 1 INVITE\r\nContent-Length: 0\r\n\r\n"}
{"level":"debug","time":"2024-10-30T08:50:26-04:00","message":"UDP write to [::]:5060 -> 127.0.0.1:5061:\nSIP/2.0 200 OK\r\nVia: SIP/2.0/UDP 127.0.0.1:5061;branch=z9hG4bK4da8.26650\r\nFrom: \"sipp\" <sip:sipp@127.0.0.1:5061>;tag=26650SIPpTag001\r\nTo: \"service\" <sip:service@127.0.0.1:5060>;tag=2ebe37d5-ec44-4d7c-9779-792d7a7512e3\r\nCall-ID: 1-26650@127.0.0.1\r\nCSeq: 1 CANCEL\r\nContent-Length: 0\r\n\r\n"}
{"level":"debug","time":"2024-10-30T08:50:26-04:00","message":"UDP read from [::]:5060 <- 127.0.0.1:5061:\nACK sip:service@127.0.0.1:5060 SIP/2.0\r\nVia: SIP/2.0/UDP 127.0.0.1:5061;branch=z9hG4bK4da8.26650\r\nFrom: sipp <sip:sipp@127.0.0.1:5061>;tag=26650SIPpTag001\r\nTo: service <sip:service@127.0.0.1:5060>;tag=8f4affcf-9ca4-4ebe-a36b-4f60f844509b\r\nCall-ID: 1-26650@127.0.0.1\r\nCSeq: 1 ACK\r\nContact: sip:sipp@127.0.0.1:5061\r\nMax-Forwards: 70\r\nSubject: Performance Test\r\nContent-Length: 0\r\n\r\n"}
{"level":"debug","caller":"transactionLayer","key":"z9hG4bK4da8.26650__127.0.0.1__5061__INVITE","input":"server_input_ack","state":"inviteStateCompleted-fm","time":"2024-10-30T08:50:26-04:00","message":"Changing transaction state"}
{"time":"2024-10-30T08:50:26.795442-04:00","level":"INFO","msg":"Got ACK","m":{"SipVersion":"SIP/2.0","Method":"ACK","Recipient":{"Scheme":"sip","Wildcard":false,"HierarhicalSlashes":false,"User":"service","Password":"","Host":"127.0.0.1","Port":5060,"UriParams":null,"Headers":null}}}
{"time":"2024-10-30T08:50:26.795574-04:00","level":"INFO","msg":"Transaction canceled","tx":{}}
{"level":"debug","caller":"transactionLayer","time":"2024-10-30T08:50:26-04:00","message":"Server transaction terminating"}
{"level":"debug","caller":"transactionLayer","tx":"z9hG4bK4da8.26650__127.0.0.1__5061__INVITE","time":"2024-10-30T08:50:26-04:00","message":"Server transaction destroyed"}
```

Also, the associated code:
```golang
func (s *sipServer) ListenAndServe(network, addr string) error {
	sip.SIPDebug = s.debug
	sip.TransactionFSMDebug = true

	s.srv.OnInvite(func(req *sip.Request, tx sip.ServerTransaction) {
		reason := "OK"
		code := sip.StatusOK
		body := []byte{}

		metrics.Metrics.SipRequestReceivedCount.WithLabelValues("INVITE").Inc()

		// Respond with 100 Trying
		res := sip.NewResponseFromRequest(req, 100, "Trying", nil)
		tx.Respond(res)

		metrics.Metrics.SipResponseSentCount.WithLabelValues("1XX").Inc()
		replyChan := make(chan reply, 1)
		cancelableCtx, cancelRequest := context.WithCancel(s.ctx)

		go func() {
			// Call INVITE handler
			replyBody, err := s.inviteHandler(cancelableCtx, req)
			if replyBody != nil && err == nil {
				body, err = json.Marshal(replyBody)
			}
			if err != nil {
				code = sip.StatusInternalServerError
				reason = "Internal Server Error"
			}

			replyChan <- reply{replyBody, err}
		}()

		select {
			case m := <-tx.Acks(): // Handle ACK for response. ACKs on 2xx are send as different request
				metrics.Metrics.SipRequestReceivedCount.WithLabelValues("ACK").Inc()
				log.Info("Got ACK", "m", m, )
				err := tx.Err()
				if errors.Is(err, sip.ErrTransactionCanceled) {
					// TODO: Find a propoer way to get CANCEL request ... discussion here https://github.com/emiago/sipgo/issues/148

					// Transaction was canceled
					metrics.Metrics.SipRequestReceivedCount.WithLabelValues("CANCEL").Inc()
					log.Info("Transaction canceled", "tx", tx)
				}
				return

			case <-tx.Done():
				// Signal transaction is done.
				// Check any errors with tx.Err() to have more info why terminated
				log.Info("Transaction done", "tx", tx)
				cancelRequest() // Signal inviteHandler to cancel
				return

			case reply := <-replyChan:
				// Handle reply from handler
				if reply.err != nil {
					log.Error("Error while handling INVITE", "error", reply.err)
				}
				log.Info("Reply from handler", "reply", reply.replyBody)

				// INVITE Final answer
				res = sip.NewResponseFromRequest(req, code, reason, body)
				tx.Respond(res)

				metricsClass := fmt.Sprintf("%dXX", int(code/100))
				metrics.Metrics.SipResponseSentCount.WithLabelValues(metricsClass).Inc()
				return
		}
	})

	s.srv.OnAck(func(req *sip.Request, tx sip.ServerTransaction) {
		metrics.Metrics.SipRequestReceivedCount.WithLabelValues("ACK").Inc()
		log.Info("Got ACK", "req", req)
	})

	return s.srv.ListenAndServe(s.ctx, network, addr)
}
```

**Comment by emiago** [comment]
Hi @SylvainFontaineJive thanks for reporting.  Couple notes, I see no reason forking invite handling, maybe you can explain that part. 

So transaction FSM is ok. 

Anyway I think for UDP transaction is put on hold due to timer.  Timer I
https://datatracker.ietf.org/doc/html/rfc3261#section-17.2.1

but yes you get ACK before. 
I think I see bug with TCP here as well, but this needs to be tested

**Comment by SylvainFontaineJive** [comment]
Hello @emiago,

I fork Invite handling because processing will take time (>300 ms).
I order to receive `tx.Acks()` (and cancel) while processing the request, I need the invite go routine to listen to those events.

Do you have an idea on how to proceed, I could submit a PR if you'd like.

**Comment by emiago** [comment]
Hi @SylvainFontaineJive lets just discuss more first. 

So just to repeat. `tx.Done()` may not fire until timer I is finished. You should have ErrTransactionCanceled after as well?
This is due to UDP mostly.  Can you confirm this please?

Now what I see you want some mechanism to Cancel current execution once CANCEL is received. We had this before, but it forced more forking. Of course it is safe to check transaction ERR.
Still to make this more intuitive I think I need to reconsider some hooks. Channels are not good.

FYI: you have freedom to make timers smaller or bigger. 

SIDE NOTE:
How this is used with Dialog management. 
You `ReadInvite` and then start sending response or doing other work. Once CANCEL hits it terminates CALL after timer and `dialog.Context` is canceled.

**Comment by SylvainFontaineJive** [comment]
Hello @emiago,

As I said in https://github.com/emiago/sipgo/issues/148, I removed the `return` statement from the `case <-tx.Acks():` which should help getting `case <-tx.Done():` ;).  My bad.

I am getting the `<-tx.Done()` correctly, after timer I like you mentioned.

Being notified about the cancel after the ACK is enough for my needs.

Thank you again @emiago for you great code and your great support.

Regards,
Sylvain

**Comment by emiago** [comment]
Hi @SylvainFontaineJive Nice to hear. 
Still I think I saw one issue with TimerI. If you can confirm but over TCP you should not have to wait timer I

**Comment by SylvainFontaineJive** [comment]
Hi @emiago 
Indeed I do see having to wait for timer I even in TCP.  Here are the logs:
```
{"level":"debug","caller":"transport<TCP>","time":"2024-11-04T10:18:25-05:00","message":"begin listening on TCP [::]:4040"}
{"level":"debug","caller":"transport<TCP>","raddr":"127.0.0.1:65004","time":"2024-11-04T10:18:31-05:00","message":"New connection"}
{"level":"debug","ip":"127.0.0.1:4040","dst":"127.0.0.1:65004","ref":2,"time":"2024-11-04T10:18:31-05:00","message":"TCP reference increment"}
{"level":"debug","ip":"127.0.0.1:4040","dst":"127.0.0.1:65004","ref":2,"time":"2024-11-04T10:18:31-05:00","message":"TCP reference increment"}
{"level":"debug","time":"2024-11-04T10:18:31-05:00","message":"TCP read 127.0.0.1:4040 <- 127.0.0.1:65004:\nINVITE sip:service@127.0.0.1:4040 SIP/2.0\r\nVia: SIP/2.0/TCP 127.0.0.1:65004;branch=z9hG4bK4da8.6469\r\nFrom: sipp <sip:sipp@127.0.0.1:65004>;tag=6469SIPpTag001\r\nTo: service <sip:service@127.0.0.1:4040>\r\nCall-ID: 1-6469@127.0.0.1\r\nCSeq: 1 INVITE\r\nContact: sip:sipp@127.0.0.1:65004\r\nX-CustomHeader: Test; param1=1; param2=2\r\nMax-Forwards: 70\r\nSubject: Performance Test\r\nContent-Type: application/sdp\r\nContent-Length:   129\r\n\r\nv=0\r\no=user1 53655765 2353687637 IN IP4 127.0.0.1\r\ns=-\r\nc=IN IP4 127.0.0.1\r\nt=0 0\r\nm=audio 6000 RTP/AVP 0\r\na=rtpmap:0 PCMU/8000\r\n"}
{"level":"debug","caller":"transportlayer","network":"tcp","addr":"127.0.0.1:65004","time":"2024-11-04T10:18:31-05:00","message":"getting connection"}
{"level":"debug","ip":"127.0.0.1:4040","dst":"127.0.0.1:65004","ref":3,"time":"2024-11-04T10:18:31-05:00","message":"TCP reference increment"}
{"level":"debug","caller":"transactionLayer","tx":"z9hG4bK4da8.6469__127.0.0.1__65004__INVITE","time":"2024-11-04T10:18:31-05:00","message":"Server transaction initialized"}
{"level":"debug","caller":"transactionLayer","key":"z9hG4bK4da8.6469__127.0.0.1__65004__INVITE","input":"server_input_user_1xx","state":"inviteStateProcceeding-fm","time":"2024-11-04T10:18:31-05:00","message":"Changing transaction state"}
{"level":"debug","time":"2024-11-04T10:18:31-05:00","message":"TCP write 127.0.0.1:4040 -> 127.0.0.1:65004:\nSIP/2.0 100 Trying\r\nVia: SIP/2.0/TCP 127.0.0.1:65004;branch=z9hG4bK4da8.6469\r\nFrom: \"sipp\" <sip:sipp@127.0.0.1:65004>;tag=6469SIPpTag001\r\nTo: \"service\" <sip:service@127.0.0.1:4040>\r\nCall-ID: 1-6469@127.0.0.1\r\nCSeq: 1 INVITE\r\nContent-Length: 0\r\n\r\n"}
{"time":"2024-11-04T10:18:31.281989-05:00","level":"INFO","msg":"HandleInvite called","call-id":"1-6469@127.0.0.1","from":"sipp","to":"service","remote":"127.0.0.1:65004"}
{"level":"debug","time":"2024-11-04T10:18:31-05:00","message":"TCP read 127.0.0.1:4040 <- 127.0.0.1:65004:\nCANCEL sip:service@127.0.0.1:4040 SIP/2.0\r\nVia: SIP/2.0/TCP 127.0.0.1:65004;branch=z9hG4bK4da8.6469\r\nFrom: sipp <sip:sipp@127.0.0.1:65004>;tag=6469SIPpTag001\r\nTo: service <sip:service@127.0.0.1:4040>\r\nCall-ID: 1-6469@127.0.0.1\r\nCSeq: 1 CANCEL\r\nContact: sip:sipp@127.0.0.1:65004\r\nMax-Forwards: 70\r\nSubject: Performance Test\r\nContent-Type: application/sdp\r\nContent-Length:     0\r\n\r\n"}
{"level":"debug","caller":"transactionLayer","key":"z9hG4bK4da8.6469__127.0.0.1__65004__INVITE","input":"server_input_cancel","state":"inviteStateProcceeding-fm","time":"2024-11-04T10:18:31-05:00","message":"Changing transaction state"}
{"level":"debug","caller":"transactionLayer","time":"2024-11-04T10:18:31-05:00","message":"Passing 487 on CANCEL"}
{"level":"debug","caller":"transactionLayer","key":"z9hG4bK4da8.6469__127.0.0.1__65004__INVITE","input":"server_input_user_300_plus","state":"inviteStateProcceeding-fm","time":"2024-11-04T10:18:31-05:00","message":"Changing transaction state"}
{"level":"debug","time":"2024-11-04T10:18:31-05:00","message":"TCP write 127.0.0.1:4040 -> 127.0.0.1:65004:\nSIP/2.0 487 Request Terminated\r\nVia: SIP/2.0/TCP 127.0.0.1:65004;branch=z9hG4bK4da8.6469\r\nFrom: \"sipp\" <sip:sipp@127.0.0.1:65004>;tag=6469SIPpTag001\r\nTo: \"service\" <sip:service@127.0.0.1:4040>;tag=5265fd4b-bd60-453c-9fd4-8b58fa29ecf1\r\nCall-ID: 1-6469@127.0.0.1\r\nCSeq: 1 INVITE\r\nContent-Length: 0\r\n\r\n"}
{"level":"debug","time":"2024-11-04T10:18:31-05:00","message":"TCP write 127.0.0.1:4040 -> 127.0.0.1:65004:\nSIP/2.0 200 OK\r\nVia: SIP/2.0/TCP 127.0.0.1:65004;branch=z9hG4bK4da8.6469\r\nFrom: \"sipp\" <sip:sipp@127.0.0.1:65004>;tag=6469SIPpTag001\r\nTo: \"service\" <sip:service@127.0.0.1:4040>;tag=97d76a7e-54a4-4a8b-935f-12aebe0d4221\r\nCall-ID: 1-6469@127.0.0.1\r\nCSeq: 1 CANCEL\r\nContent-Length: 0\r\n\r\n"}
{"level":"debug","time":"2024-11-04T10:18:31-05:00","message":"TCP read 127.0.0.1:4040 <- 127.0.0.1:65004:\nACK sip:service@127.0.0.1:4040 SIP/2.0\r\nVia: SIP/2.0/TCP 127.0.0.1:65004;branch=z9hG4bK4da8.6469\r\nFrom: sipp <sip:sipp@127.0.0.1:65004>;tag=6469SIPpTag001\r\nTo: service <sip:service@127.0.0.1:4040>;tag=5265fd4b-bd60-453c-9fd4-8b58fa29ecf1\r\nCall-ID: 1-6469@127.0.0.1\r\nCSeq: 1 ACK\r\nContact: sip:sipp@127.0.0.1:65004\r\nMax-Forwards: 70\r\nSubject: Performance Test\r\nContent-Length: 0\r\n\r\n"}
{"level":"debug","time":"2024-11-04T10:18:31-05:00","message":"TCP read 127.0.0.1:4040 <- 127.0.0.1:65004:\n"}
{"level":"debug","caller":"transport<TCP>","error":"EOF","time":"2024-11-04T10:18:31-05:00","message":"connection was closed"}
{"level":"debug","ip":"127.0.0.1:4040","dst":"127.0.0.1:65004","ref":2,"time":"2024-11-04T10:18:31-05:00","message":"TCP reference decrement"}
{"level":"debug","ip":"127.0.0.1:4040","dst":"127.0.0.1:65004","ref":0,"time":"2024-11-04T10:18:31-05:00","message":"TCP doing hard close"}
{"level":"debug","caller":"transactionLayer","key":"z9hG4bK4da8.6469__127.0.0.1__65004__INVITE","input":"server_input_ack","state":"inviteStateCompleted-fm","time":"2024-11-04T10:18:31-05:00","message":"Changing transaction state"}
{"time":"2024-11-04T10:18:31.385218-05:00","level":"INFO","msg":"Transaction canceled","call-id":"1-6469@127.0.0.1","from":"sipp","to":"service","remote":"127.0.0.1:65004"}
```
Wait time then:
```
{"level":"debug","caller":"transactionLayer","key":"z9hG4bK4da8.6469__127.0.0.1__65004__INVITE","input":"server_input_timer_i","state":"inviteStateConfirmed-fm","time":"2024-11-04T10:18:36-05:00","message":"Changing transaction state"}
{"level":"debug","caller":"transactionLayer","tx":"z9hG4bK4da8.6469__127.0.0.1__65004__INVITE","time":"2024-11-04T10:18:36-05:00","message":"Server transaction destroyed"}
{"level":"debug","caller":"transactionLayer","time":"2024-11-04T10:18:36-05:00","message":"Server transaction terminating"}
{"level":"debug","caller":"transactionLayer","tx":"z9hG4bK4da8.6469__127.0.0.1__65004__INVITE","time":"2024-11-04T10:18:36-05:00","message":"Server transaction destroyed"}
{"time":"2024-11-04T10:18:36.386463-05:00","level":"INFO","msg":"Transaction Done","call-id":"1-6469@127.0.0.1","from":"sipp","to":"service","remote":"127.0.0.1:65004"}

```

**Comment by emiago** [comment]
@SylvainFontaineJive it is fixed. You can test in main branch
